aten/src/ATen/native/cpu/BinaryOpsKernel.cpp

```diff
@@ -313,6 +313,14 @@ void bitwise_xor_kernel(TensorIteratorBase& iter) {
 }
 
 void lshift_kernel(TensorIteratorBase& iter) {
+#if defined(__VSX__)  || defined(CPU_CAPABILITY_VSX)
+  AT_DISPATCH_INTEGRAL_TYPES(iter.dtype(), "lshift_cpu", [&]() {
+    cpu_kernel(iter,
+      [](scalar_t a, scalar_t b) -> scalar_t {
+        return static_cast<std::make_unsigned_t<scalar_t>>(a) << b;
+    });
+  });
```

Comment:
Is there vectorized bit shift support in VSX to add in `vec256/vsx/` so that we don't have to exclude it explicitly here in the common aten kernel?

Reply:
@jgong5, I am not sure on it. But this presently fixes the build errors that we are seeing. 
@quickwritereader, any idea if we have bit shift support in VSX?

Reply:
I believe left shift should be this one https://www.ibm.com/docs/en/xl-c-and-cpp-linux/16.1.0?topic=functions-vec-sldw
but still, it's worth defining if it's not defined. similarly this way inside here https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/cpu/vec/vec256/vsx/vsx_helpers.h#L71-L83
but with c as an argument for template type T

Reply:
something like this for 4 byte:
vec_sl( vec , vec_splats(static_cast<unsigned int>(shift_int) ));

Reply:
so it seems its better to write that code for each type or to use macros. I do not know why I wrote it macros heavily instead of template heavily like I did zarch version.

Reply:
Thanks @quickwritereader, I will try this.

Reply:
@quickwritereader, I tried using the following in `vsx_helpers.h` along with the respective macro calls in the associated header files. 

```
#define DEFINE_MEMBER_OP2(op, op_type, func)                                  \
  Vectorized<op_type> C10_ALWAYS_INLINE operator<<(unsigned int shift) const {
     auto shift_vec = vec_splats( shift );
    return {vec_sl(_vec0, shift_vec), vec_sl(_vec1,shift_vec )};
   }


```
but I get errors of the kind:

```
<mydir>work/aten/src/ATen/cpu/vec/vec256/vsx/vsx_helpers.h:177:35: error: 'shift' was not declared in this scope
<mydir>work/aten/src/ATen/cpu/vec/vec256/vsx/vsx_helpers.h:177:41: error: invalid parameter combination for AltiVec intrinsic '__builtin_vec_splats'
<mydir>work/aten/src/ATen/cpu/vec/vec256/vsx/vsx_helpers.h:178:5: error: expected unqualified-id before 'return'
<mydir>work/aten/src/ATen/cpu/vec/vec256/vsx/vsx_helpers.h:179:4: error: expected declaration before '}' token
<mydir>work/aten/src/ATen/cpu/vec/vec256/vsx/vsx_helpers.h:159:13: error: invalid parameter combination for AltiVec intrinsic '__builtin_vec_sl'
<mydir>work/aten/src/ATen/cpu/vec/vec256/vsx/vsx_helpers.h:159:39: error: invalid parameter combination for AltiVec intrinsic '__builtin_vec_sl'
<mydir>work/aten/src/ATen/cpu/vec/vec256/vsx/vsx_helpers.h:159:59: error: no matching function for call to 'at::vec::DEFAULT::Vectorized<short int>::Vectorized(<brace-enclosed initializer list>)'
<mydir>work/aten/src/ATen/cpu/vec/vec256/vsx/vsx_helpers.h:159:13: error: invalid parameter combination for AltiVec intrinsic '__builtin_vec_sl'
<mydir>work/aten/src/ATen/cpu/vec/vec256/vsx/vsx_helpers.h:159:39: error: invalid parameter combination for AltiVec intrinsic '__builtin_vec_sl'
<mydir>work/aten/src/ATen/cpu/vec/vec256/vsx/vsx_helpers.h:159:59: error: no matching function for call to 'at::vec::DEFAULT::Vectorized<int>::Vectorized(<brace-enclosed initializer list>)'
<mydir>work/aten/src/ATen/cpu/vec/vec256/vsx/vsx_helpers.h:159:13: error: invalid parameter combination for AltiVec intrinsic '__builtin_vec_sl'
<mydir>work/aten/src/ATen/cpu/vec/vec256/vsx/vsx_helpers.h:159:39: error: invalid parameter combination for AltiVec intrinsic '__builtin_vec_sl'
<mydir>work/aten/src/ATen/cpu/vec/vec256/vsx/vsx_helpers.h:159:59: error: no matching function for call to 'at::vec::DEFAULT::Vectorized<long int>::Vectorized(<brace-enclosed initializer list>)'
```
Do you think I am missing something obvious?

Reply:
I see.
I think type should match. See if (op_type shift) works

Reply:
And also imho better rename it as DEFINE_SHIFT_LEFT,  as it is used specially for that operation

Reply:
yes, I just pasted trial code, yet to update the name of the macro. I will update it in final code.